Fonts-weight descriptor (#36686)
mdn/content: files/en-us/web/css/@font-face/font-weight/index.md

{{CSSRef}}

The **`font-weight`** CSS {{cssxref("@font-face")}} descriptor enables authors to specify a single font weight, or a range of font weights, for the font specified in a {{cssxref("@font-face")}} at-rule. This is then used by the browser to select the appropriate font when a CSS rule sets a desired {{cssxref("font-weight", "font weight")}}.

Unless it contains a [variable font](/en-US/docs/Web/CSS/CSS_fonts/Variable_fonts_guide), a single font file contains characters from a font family in a specific weight and style: for example, "Helvetica bold italic". To declare the font to be used for a range of font weights, declare a space-separated pair of font-weight values as the value for the `font-weight` descriptor.

> Typically, a developer will want to use fonts from a single font family in a range of different weights. To accomplish this, you can define multiple {{cssxref("@font-face")}} at-rules for the same family (all with the same {{cssxref("@font-face/font-family", "font-family")}} descriptor value), one for each weight, and set the `font-weight` descriptor to define the weight range for which that particular font file will be used.

Author: Thanks for the review. I rewrote the intro so i think its a bit better than my mishmash i had before.

After this, CSS rules can select the normal or the bold font for the "Fira Sans" family just by setting the {{cssxref("font-weight")}} property. Note that the {{htmlelement("strong")}} HTML element also selects the bold font, because by default `<strong>` elements have a CSS `font-weight` property value of `bold`.

#### HTML

```html
<p class="one">Fira Sans, `normal` weight paragraph</p>
<p class="two">Fira Sans, `bold` weight paragraph</p>
<p><strong>Fira Sans, &lt;strong&gt; element (`bold`)</strong></p>
```

#### CSS

```css
@font-face {
  font-family: "Fira Sans";
  font-weight: normal;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-Regular.woff2");
}

@font-face {
  font-family: "Fira Sans";
  font-weight: bold;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-Bold.woff2");
}

body {
  font-family: "Fira Sans", serif;
  font-size: 2rem;
}

p.one {
  font-weight: normal;
}

p.two {
  font-weight: bold;
}
```

#### Result

{{embedlivesample("Selecting normal and bold fonts", "", 300)}}

> ### Setting font-weight ranges

Reviewer: I don't understand the point of this example. Why would someone want to use a range with a non-variable font?

ISTR when I looked into this the only use case I could see for a range was to limit the weights available in a variable range font.

Author: i've actually seen this done in the wild. <s>redacted</s> had 6 different font names for the same font in different ranges, and they updated the font-family name for every font-weight property. People do it, so we need to demonstrate it.

Reviewer: But what's the use case for it? As someone trying to learn this feature, I look at this and think, why would anyone want to do this, what does it accomplish?

> People do it, so we need to demonstrate it.

I'm not sure I agree with this. People do all kinds of things, but we should demo sensible use cases.

Reviewer: Also, this:

> redacted had 6 different font names for the same font in different ranges, and they updated the font-family name for every font-weight property. 

...isn't what's happening here, is it? This has the same font-family name.

Author: no; here we are using one font family name. I'll add to the pre-explanation.

> In this example, we include the same two fonts as in the example above, but we set the `font-weight` descriptors to a different ranges.

Author: ```suggestion
In this example, we include the same two fonts as in the example above, but we set the `font-weight` descriptors to a different ranges. 

While this example is a bit contrived, it demonstrates how authors can include multiple fonts for multiple font-weights (and font-styles), by including multiple `@font-face` declarations with the same `font-family` value. In the rest of your stylesheets, you can then declare a `font-weight` (or `font-style`), and know the appropriate font will be used.
```

#### HTML

```html
<p class="one">Fira Sans, `normal` weight paragraph</p>
<p class="two">Fira Sans, `bold` weight paragraph</p>
<p><strong>Fira Sans, &lt;strong&gt; element (`bold`)</strong></p>
```

#### CSS

```css
@font-face {
  font-family: "Fira Sans";
  font-weight: normal;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-Regular.woff2");
}

@font-face {
  font-family: "Fira Sans";
  font-weight: bold;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-Bold.woff2");
}

body {
  font-family: "Fira Sans", serif;
  font-size: 2rem;
}

p.one {
  font-weight: normal;
}

p.two {
  font-weight: bold;
}
```

#### Result

{{embedlivesample("Selecting normal and bold fonts", "", 300)}}

### Setting font-weight ranges

In this example, we include the same two fonts as in the example above, but we set the `font-weight` descriptors to a different ranges.

> While this example is a bit contrived, it demonstrates how authors can include multiple fonts for multiple font-weights (and font-styles), by including multiple `@font-face` declarations with the same `font-family` value. In the rest of your stylesheets, you can then declare a `font-weight` (or `font-style`), and know the appropriate font will be used.

Reviewer: Sorry if I'm being thick, but the previous examples demonstrates this use case as well, doesn't it? (and is not contrived, AFAIK it is showing the main use case for `font-weight`). So what is the point here of using a range?

Reviewer: OK, I think that makes sense! But then I still don't understand why the second range is a subset of the first. Why do we want to select the non-bold font when the dev writes `1000`?

I also wondered if you could do this without ranges, like this:

```css
@font-face {
  font-family: "Fira Sans";
  font-weight: 1;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-Light.woff2");
}
@font-face {
  font-family: "Fira Sans";
  font-weight: 200;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-Regular.woff2");
}

@font-face {
  font-family: "Fira Sans";
  font-weight: 400;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-Bold.woff2");
}

@font-face {
  font-family: "Fira Sans";
  font-weight: 900;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-ExtraBold.woff2");
}
```

What do ranges give you over this?

Reviewer: It looks like this^ gives some results that I would not (naively) expect, in particular `font-weight: 600;
` selected extrabold.

You do need something like this:

```css
@font-face {
  font-family: "Fira Sans";
  font-weight: 1;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-Light.woff2");
}

@font-face {
  font-family: "Fira Sans";
  font-weight: 200;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-Regular.woff2");
}

@font-face {
  font-family: "Fira Sans";
  font-weight: 400;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-Bold.woff2");
}

@font-face {
  font-family: "Fira Sans";
  font-weight: 900 1000;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-ExtraBold.woff2");
}
```

...to get the results I was expecting. I don't know why though.

Reviewer: It looks like this^ gives some results that I would not (naively) expect, in particular `font-weight: 600;
` selected extrabold.

With something like this:

```css
@font-face {
  font-family: "Fira Sans";
  font-weight: 1 199;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-Light.woff2");
}

@font-face {
  font-family: "Fira Sans";
  font-weight: 200 399;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-Regular.woff2");
}

@font-face {
  font-family: "Fira Sans";
  font-weight: 400 899;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-Bold.woff2");
}

@font-face {
  font-family: "Fira Sans";
  font-weight: 900;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-ExtraBold.woff2");
}
```

... I get the results I was expecting. I don't know why though.

Author: i used something similar to your example; the ranges are slightly broader so i can explain the cascade (and so we don't need to clarify 399.5)

Reviewer: Thanks! This makes sense to me. I think I would still prefer it without the cascade demo (i.e. presenting each range "in order"), but this is OK for me.

}

p.five {
  font-weight: 500;
}

p.seven {
  font-weight: 700;
}

p.nine {
  font-weight: 900;
}
```

#### Result

{{embedlivesample("Setting font-weight ranges", "", 300)}}

The `seven` paragraph uses the extra bold font. While `font-weight: 700` matches both the FiraSans-Bold and FiraSans-ExtraBold declarations, as the FiraSans-ExtraBold is declared later, it overrides the FiraSans-Bold for that value. Similarly, the `100` and `300` both use the FiraSans-Light; which FiraSans-Regular and FiraSans-Light both include `300` in their ranges, FiraSans-Light is declared later. Alternatively, we could have declared FiraSans-Regular after FiraSans-Light, but we would need to change the `font-weight` descriptor range if we do so.

### Setting a range for a variable font

In this example, you can change the font size of the variable font. We include ["League Mono"](https://www.theleagueofmoveabletype.com/league-mono), using a single `@font-face` at-rule. We've used a `font-weight: 300 700` value which effectively limits the range of weights that are available.
When the `font-weight` property is specifies a weight outside this range the weight it gets is clamped to the range.

#### HTML

> We include a paragraph with `<output>` initially set to `400`, as that is the the default font-weight for unstyled paragraph text. This paragraph is nestled between two other paragraphs to enable comparing rendered versus declared font weight values.

Reviewer: ```suggestion
We include a paragraph with `<output>` initially set to `400`, as that is the the default font-weight for unstyled paragraph text. This paragraph is nestled between two other paragraphs, so you can compare rendered versus declared font weight values.
```

- the bold font is associated with the range 500-700
- the extra-bold font is associated with the range 700-1000

The CSS [cascade](/en-US/docs/Web/CSS/Cascade) ensures that the three latter declarations override parts of the the range that was set in the `FiraSans-Regular` declaration.

```css
@font-face {
  font-family: "Fira Sans";
  font-weight: 1 1000;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-Regular.woff2");
}

@font-face {
  font-family: "Fira Sans";
  font-weight: 1 300;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-Light.woff2");
}

@font-face {
  font-family: "Fira Sans";
  font-weight: 500 700;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-Bold.woff2");
}

@font-face {
  font-family: "Fira Sans";
  font-weight: 700 1000;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-ExtraBold.woff2");
}

body {
  font-family: "Fira Sans", serif;
  font-size: 2rem;
}

p.one {
  font-weight: 100;
}

p.three {
  font-weight: 300;
}

p.four {
  font-weight: 400;
}

p.five {
  font-weight: 500;
}

p.seven {
  font-weight: 700;
}

p.nine {
  font-weight: 900;
}
```

#### Result

> {{embedlivesample("Setting font-weight ranges", "", 400)}}

Reviewer: ```suggestion
{{embedlivesample("Setting font-weight ranges", "", 500)}}
```
